src/services/safe_border.py

```diff
@@ -0,0 +1,179 @@
+        withdrawable_epoch = min(get_validators_withdrawable_epochs(validators))
+
+        last_finalized_request_id_slot = self._get_last_finalized_withdrawal_request_slot(blockstamp)
+
+        start_slot = max(last_finalized_request_id_slot, self.get_epoch_first_slot(self._get_validators_earliest_activation_epoch(validators)))
+        end_slot = min(blockstamp.ref_slot, self.get_epoch_first_slot(withdrawable_epoch - EPOCHS_PER_SLASHINGS_VECTOR))
+
+        while self._check_slot_diff(start_slot, end_slot):
+            mid_slot = (end_slot + start_slot) // 2
+            validators = self._get_archive_lido_validators_by_keys(mid_slot, pubkeys)
+            slashed_validators = filter_slashed_validators(validators)
+
+            if len(slashed_validators) > 0:
+                end_slot = mid_slot - 1
+                pubkeys = self.get_validators_pubkeysget_validators_pubkeys(slashed_validators)
+            else:
+                start_slot = mid_slot + 1
+
+        return self.get_epoch_by_slot(start_slot)
+
+    def _check_slot_diff(self, start_slot: int, end_slot: int):
+        if (end_slot - start_slot > self.chain_config.slots_per_epoch):
+            return False
+        return True
+
+    def _filter_validators_with_earliest_exit_epoch(self, validators) -> list[Validator]:
+        if len(validators) == 0:
+            return []
+
+        sorted_validators = sorted(validators, key = lambda validator: (int(validator.validator.exit_epoch)))
+        earliest_exit_epoch = sorted_validators[0].validator.exit_epoch
+        return filter_validators_by_exit_epoch(sorted_validators, earliest_exit_epoch)
+
+    def _get_validators_earliest_activation_epoch(self, validators) -> int:
+        if len(validators) == 0:
+            return 0
+
+        sorted_validators = sorted(validators, key = lambda validator: (int(validator.validator.validator.activation_epoch)))
+        return sorted_validators[0].validator.validator.activation_epoch
+
+    def _get_lido_validators(self, blockstamp: BlockStamp) -> list[Validator]:
+        return [lv.validator for lv in self.w3.lido_validators.get_lido_validators(blockstamp.ref_slot)]
+
+    def _get_archive_lido_validators_by_keys(self, ref_slot, pubkeys) -> list[Validator]:
+        return self.w3.cc.get_validators(ref_slot, tuple(pubkeys))
+
+    def _get_bunker_mode_start_timestamp(self, blockstamp: BlockStamp) -> int:
+        return self.w3.lido_contracts.withdrawal_queue.functions.bunkerModeSinceTimestamp().call(block_identifier=blockstamp.block_hash)
+
+    def _get_last_finalized_withdrawal_request_slot(self, blockstamp: BlockStamp) -> int:
+        last_finalized_request_id = self.w3.lido_contracts.withdrawal_queue.functions.getLastFinalizedRequestId().call(block_identifier=blockstamp.block_hash)
+        last_finalized_request_data = self.w3.lido_contracts.withdrawal_queue.functions.getWithdrawalRequestStatus(last_finalized_request_id).call(block_identifier=blockstamp.block_hash)
```

Comment:
need to handle the case when `last_finalized_request_id == 0` since `getWithdrawalRequestStatus` will revert